### DISCUSSION TITLE
EdrumulusGui.py idea to improve

### DISCUSSION THREAD

**Discussion by ocimjr** [corrados/edrumulus#127] [Ideas]
Hi,

I don't not it is viable or not, but when using edrumulus you have first be sure that you create EdrumulusIn and Out port before start this program. And in mac you have also to use another program to convert serial to midi messages like hairless midi etc..

Looking for a replacement on mac I found this, https://github.com/RuudMulder/SerialMidiBridge/blob/main/SerialMidiBridge.py
So, you can create midi in and out port and also connect to serial to use edrumulus 

Instead of having 2 programs, and since edrumulus use python, we could only start edrumulusGui and it would take charge to create ports needed and serial connection.

Thanks.

**Comment by corrados** [comment]
Good idea, actually. I wonder if such an implementation would work on all OSs, i.e., Windows, Linux and Mac. I'll have a look at it sometime soon.

**Reply by corrados** [reply]
It's unfortunately not so easy. If you create a serial connection to Edrumulus, it is blocked for other MIDI tools. So, it will be possible that the Edrumulus GUI talks to the Edrumulus hardware but your DAW also needs a MIDI connection to the Edrumulus hardware.

**Reply by ocimjr** [reply]
> You can create a midi port so in your daw you can select this new port

In fact thinking, you could see if there is already a port with Edrumulus In or Out, if so you connect if not you create

**Reply by corrados** [reply]
There is an experimental implementation available on new branch [esp32_direct_serial_connection](https://github.com/corrados/edrumulus/tree/esp32_direct_serial_connection).

You need to modify the correct device name (I only tested it under Linux because my Mac is too old):
https://github.com/corrados/edrumulus/blob/b3fcd30a1517faf869f95a01f16d977143f8fcd9/tools/edrumulus_gui.py#L667

Start with:
`edrumulus_gui.py serial`

**Reply by corrados** [reply]
FYI, I just merge the changes to the main branch.

**Reply by ocimjr** [reply]
Cool, thanks l will test this weekend.

**Reply by corrados** [reply]
I succeeded in running it on my old Mac. It worked with
`./edrumulus_gui.py serial /dev/tty.SLAB_USBtoUART`
using the latest code on main branch.

**Comment by corrados** [comment]
@thijstriemstra I just implemented something which will be very interesting for you. It is currently on my branch [use_pytemidi](https://github.com/corrados/edrumulus/tree/use_pytemidi) but it will be merged to main, soon.

You can now run: `python edrumulus_gui.py serial COM7` on Windows and you do not need to run any other tool like loopMIDI or Hairless MIDI anymore:
![grafik](https://github.com/corrados/edrumulus/assets/46655886/b1c5fd86-39d4-4ad1-a5ee-4c2095f1c9bb)

**Reply by corrados** [reply]
It's on main branch now. use_pytemidi is deleted.

**Reply by corrados** [reply]
If it works for you as well, we might have to re-write the following section in the Manual: https://github.com/corrados/edrumulus/blob/main/doc/manual.md#setup-for-esp32-under-windows
Most of the description can then be removed...

**Comment by corrados** [comment]
I just changed the default for edrumulus_gui.py so that it should work with the direct serial connection on each platform (Windows, Linux, Mac) without having to use any command line argument.

**Reply by thijstriemstra** [reply]
Nice! could you also document the various connection options in https://github.com/corrados/edrumulus/blob/main/doc/manual.md so users don't have to inspect the source code to check the various options?

**Reply by corrados** [reply]
Any volunteers? ;-)

**Reply by corrados** [reply]
BTW, can you please give some feedback if it works for you? Could you successfully connect your DAW to the new edrumulus_gui.py virtual MIDI port?

**Reply by ocimjr** [reply]
For me it's all working as should. Thank you

**Reply by thijstriemstra** [reply]
> BTW, can you please give some feedback if it works for you? 

For sure! TBC
